Mathlib/NumberTheory/ModularForms/JacobiTheta/TwoVariable.lean

```diff
@@ -14,119 +15,501 @@ This file defines the two-variable Jacobi theta function
+    exact Int.one_le_abs hn
+  · refine fun hτ ↦ ((summable_pow_mul_jacobiTheta₂_term_bound
+      |z.im| hτ 1).mul_left (2 * π)).of_norm_bounded _ (fun n ↦ ?_)
+    rw [jacobiTheta₂'_term, norm_mul, ← mul_assoc, pow_one]
+    refine mul_le_mul (le_of_eq ?_) (norm_jacobiTheta₂_term_le hτ le_rfl le_rfl n)
+      (norm_nonneg _) (by positivity)
+    simp_rw [norm_mul, Complex.norm_eq_abs, Complex.abs_two, abs_I,
+      Complex.abs_of_nonneg pi_pos.le, abs_intCast, mul_one, Int.cast_abs]
+
+end term_bounds
+
+/-!
+## Definitions of the functions
+-/
+
+/-- The two-variable Jacobi theta function,
+`θ z τ = ∑' (n : ℤ), cexp (2 * π * I * n * z + π * I * n ^ 2 * τ)`.
+-/
+def jacobiTheta₂ (z τ : ℂ) : ℂ := ∑' n : ℤ, jacobiTheta₂_term n z τ
+
+/-- Fréchet derivative of the two-variable Jacobi theta function. -/
+def jacobiTheta₂_fderiv (z τ : ℂ) : ℂ × ℂ →L[ℂ] ℂ := ∑' n : ℤ, jacobiTheta₂_term_fderiv n z τ
+
+/-- The `z`-derivative of the Jacobi theta function,
+`θ' z τ = ∑' (n : ℤ), 2 * π * I * n * cexp (2 * π * I * n * z + π * I * n ^ 2 * τ)`.
+ -/
+def jacobiTheta₂' (z τ : ℂ) := ∑' n : ℤ, jacobiTheta₂'_term n z τ
+
+lemma hasSum_jacobiTheta₂_term (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    HasSum (fun n ↦ jacobiTheta₂_term n z τ) (jacobiTheta₂ z τ) :=
+  ((summable_jacobiTheta₂_term_iff z τ).mpr hτ).hasSum
+
+lemma hasSum_jacobiTheta₂_term_fderiv (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    HasSum (fun n ↦ jacobiTheta₂_term_fderiv n z τ) (jacobiTheta₂_fderiv z τ) :=
+  ((summable_jacobiTheta₂_term_fderiv_iff z τ).mpr hτ).hasSum
+
+lemma hasSum_jacobiTheta₂'_term (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    HasSum (fun n ↦ jacobiTheta₂'_term n z τ) (jacobiTheta₂' z τ) :=
+  ((summable_jacobiTheta₂'_term_iff z τ).mpr hτ).hasSum
+
+lemma jacobiTheta₂_undef (z : ℂ) {τ : ℂ} (hτ : im τ ≤ 0) : jacobiTheta₂ z τ = 0 := by
+  apply tsum_eq_zero_of_not_summable
+  rw [summable_jacobiTheta₂_term_iff]
+  exact not_lt.mpr hτ
+
+lemma jacobiTheta₂_fderiv_undef (z : ℂ) {τ : ℂ} (hτ : im τ ≤ 0) : jacobiTheta₂_fderiv z τ = 0 := by
+  apply tsum_eq_zero_of_not_summable
+  rw [summable_jacobiTheta₂_term_fderiv_iff]
+  exact not_lt.mpr hτ
+
+lemma jacobiTheta₂'_undef (z : ℂ) {τ : ℂ} (hτ : im τ ≤ 0) : jacobiTheta₂' z τ = 0 := by
+  apply tsum_eq_zero_of_not_summable
+  rw [summable_jacobiTheta₂'_term_iff]
+  exact not_lt.mpr hτ
+
+/-!
+## Derivatives and continuity
+-/
+
+lemma hasFDerivAt_jacobiTheta₂ (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    HasFDerivAt (fun p : ℂ × ℂ ↦ jacobiTheta₂ p.1 p.2) (jacobiTheta₂_fderiv z τ) (z, τ) := by
+  obtain ⟨T, hT, hτ'⟩ := exists_between hτ
+  obtain ⟨S, hz⟩ := exists_gt |im z|
+  let V := {u | |im u| < S} ×ˢ {v | T < im v}
+  have hVo : IsOpen V := by
+    refine ((_root_.continuous_abs.comp continuous_im).isOpen_preimage _ isOpen_Iio).prod ?_
+    exact continuous_im.isOpen_preimage _ isOpen_Ioi
+  have hVmem : (z, τ) ∈ V := ⟨hz, hτ'⟩
+  have hVp : IsPreconnected V := by
+    refine (Convex.prod ?_ (convex_halfspace_im_gt T)).isPreconnected
```

Comment:
This line costs 60000 heartbeats (!)
It looks like trying shortcut instances could really speed things up...

Reply:
For some reason, it is vastly quicker (several orders of magnitude) to use `IsPreconnected.prod` and then apply `Convex.isPreconnected` to the factors.